mavros/src/plugins/sys_status.cpp

```diff
@@ -920,16 +925,64 @@ class SystemStatusPlugin : public plugin::PluginBase
 		batt_msg->present = true;
 
 		batt_msg->cell_voltage.clear();
-		batt_msg->cell_voltage.reserve(bs.voltages.size());
+		batt_msg->cell_voltage.reserve(bs.voltages.size() + bs.voltages_ext.size());
+		float cell_voltage;
+		float voltage_acc = 0.0f;
+		float total_voltage = 0.0f;
+		int nr_cells = 0;
 		for (auto v : bs.voltages) {
 			if (v == UINT16_MAX)
 				break;
 
-			batt_msg->cell_voltage.push_back(v / 1000.0f);	// 1 mV
+			if (v == UINT16_MAX-1) // cell voltage is above 65,534V, add the next array element to get the correct voltage
+			{
+				voltage_acc += v * 0.001f;
+				continue;
+			}
+			else
+			{
+				voltage_acc = 0.0f;
+			}
+			cell_voltage = voltage_acc + (v * 0.001f);	// 1 mV
+			batt_msg->cell_voltage.push_back(cell_voltage);
+			total_voltage += cell_voltage;
+			nr_cells++;
 		}
+		for (auto v : bs.voltages_ext) {
+			if (v == UINT16_MAX || v == 0)  // this one is different from the for loop above to support mavlink2 message truncation
+				break;
+
+			if (v == UINT16_MAX-1) // cell voltage is above 65,534V, add the next array element to get the correct voltage
+			{
+				voltage_acc += v * 0.001f;
+				continue;
+			}
+			else
+			{
+				voltage_acc = 0.0f;
+			}
+			cell_voltage = voltage_acc + (v * 0.001f);	// 1 mV
+			batt_msg->cell_voltage.push_back(cell_voltage);
+			total_voltage += cell_voltage;
+			nr_cells++;
+		}
+		batt_msg->voltage = total_voltage;
+		batt_msg->cell_voltage.erase(batt_msg->cell_voltage.begin()+nr_cells, batt_msg->cell_voltage.end()); // erase the unused cells
```

Comment:
That's unclear to me. Why do you need `nr_cells`? Why not to use `cell_voltage.size()`?
And why do you need to `erase()`? [`reserve()`][1] do not change `size()`, only `capacity()`.

[1]: https://en.cppreference.com/w/cpp/container/vector/reserve

Reply:
> Why do you need nr_cells? Why not to use cell_voltage.size()?
because if for some reason a cell has more than 65,534V it get coalesced with the next one. Hence the bookkeeping with the extra variable.

> And why do you need to erase()?
Using erase allows up to neatly only present the actual number of cells in the ROS message.

> reserve do not change size(), only capacity().
Yes I an aware of that. This method gets a neat result with very low complexity. Erase will not realloc, nor move data around.

> processing voltages and voltages_ext might be done in lambda, so you could remove duplicate.
The hadling in voltages and voltages_ext is different. The duplication is required. :(

Reply:
Sorry, you are 100% correct about using .size() and removing nr_cells

I changed it now, and corrected other small innacuracy

Reply:
`nr_cell` always will be equal to `size()` because it's increase done after push_back().
Lambda could look like that:
```
auto copy_cells = [&](uint16_t v, bool is_ext) {
  if (v == UINT16_MAX || (is_ext && v == 0))
    return;
    
  if (v == UINT18_MAX-1) {
    volt_acc = v;
    return;
  }
  
  auto cv = (static_cast<int32_t>(volt_acc) + v) * 0.001f;
  volt_acc = 0;
  batt_msg->cell_voltage.push_back(cv);
  total_volt += cv;
}

for (auto v : bs.voltages) copy_cell(v, false);
for (auto v : bs.voltages_ext) copy_cell(v, true);

Reply:
And about ROS message: you don't have to worry about erasing capacity(), only size() matters.

Reply:
// i'm curious, that chemistry type has so huge cell voltage?

Reply:
yes, you are correct the erase is not needed at all. Removed it, sorry.

Reply:
I'm not sure about chemistry but if you have a > 16S battery, you will need that mechanism to correctly get the battery voltage. According to the mavlink standard:
https://mavlink.io/en/messages/common.html#BATTERY_STATUS

Reply:
Ah, yes, so that a hack to send above 65 total volts. I changed lambda to be able to handle more than one (2**16-2) "cells".
Please note that your code has error in `else {acc=0}` block because you're clearing accumulator before use.